examples/tutorials/matmul/int4_weight_decompression.cpp

```diff
@@ -0,0 +1,265 @@
+    // - mask = (1 << 1) would apply zero points only along the N dimension.
+    int mask = (1 << 0) + (1 << 1); // zero points both along K and N dimensions
+    memory::dims groups = {};
+    attr.set_zero_points(DNNL_ARG_WEIGHTS, mask, groups, memory::data_type::s4);
+
+    // Set fpmath mode with `apply_to_int=true` to apply fpmath mode behavior to
+    // integral primitives (in this example, matmul).
+    attr.set_fpmath_mode(fpmath_mode::f16, true);
+
+    // Create a MatMul primitive descriptor
+    return matmul::primitive_desc(eng, a_md, b_md, c_md, attr);
+}
+
+// Function to perform matrix multiplication with int4 weights decompression
+// using oneDNN
+void weights_decompression_matmul(int64_t M, int64_t N, int64_t K, int64_t G,
+        std::vector<float> &A_f32, std::vector<float> &B_f32,
+        std::vector<float> &zp_B_f32, std::vector<float> &sc_B,
+        std::vector<float> &C_f32, const engine &eng) {
+    auto matmul_pd = matmul_pd_create(M, N, K, G, eng);
+    stream s(eng);
+
+    // Pre-packed weights stored as int4
+    memory B_s4_mem(matmul_pd.weights_desc(), eng);
+    {
+        memory B_f32_mem(
+                {{K, N}, memory::data_type::f32, memory::format_tag::ab}, eng);
+        write_to_dnnl_memory(B_f32.data(), B_f32_mem);
+        reorder(B_f32_mem, B_s4_mem).execute(s, B_f32_mem, B_s4_mem);
+        s.wait();
+    }
+    matmul matmul_p(matmul_pd);
+
+    // input of the current layer / operation
+    memory A_f32_mem({{M, K}, memory::data_type::f32, {K, 1}}, eng);
+    // De-quantization parameters (eg. Scale and Shift)
+    const int64_t n_groups = K / G;
+    memory sc_B_mem({{N, n_groups}, memory::data_type::f32, {1, N}}, eng);
+
+    // Pre-packed zp stored as int4
+    // A unique zero point is used for each weight in this example
+    // Allocates memory for zp_B_s4_mem with specified dimensions and data type.
+    memory zp_B_s4_mem({{K, N}, memory::data_type::s4, {1, K}}, eng);
+    {
+        memory zp_B_f32_mem({{K, N}, memory::data_type::f32, {1, K}}, eng);
+        write_to_dnnl_memory(zp_B_f32.data(), zp_B_f32_mem);
+        reorder(zp_B_f32_mem, zp_B_s4_mem)
+                .execute(s, zp_B_f32_mem, zp_B_s4_mem);
+        s.wait();
+    }
+
+    write_to_dnnl_memory(A_f32.data(), A_f32_mem);
+    write_to_dnnl_memory(sc_B.data(), sc_B_mem);
+
+    // output - no initialization required
+    memory C_f32_mem({{M, N}, memory::data_type::f32, {N, 1}}, eng);
+
+    matmul_p.execute(s,
+            {{DNNL_ARG_SRC, A_f32_mem}, {DNNL_ARG_WEIGHTS, B_s4_mem},
+                    {DNNL_ARG_DST, C_f32_mem},
+                    {DNNL_ARG_ATTR_SCALES | DNNL_ARG_WEIGHTS, sc_B_mem},
+                    {DNNL_ARG_ATTR_ZERO_POINTS | DNNL_ARG_WEIGHTS,
+                            zp_B_s4_mem}});
+    s.wait();
+}
+
+// Compares the results of reference matrix multiplication and oneDNN weights
+// decompression.
+void compare_ref_and_weights_decompression(engine::kind engine_kind) {
```

Comment:
It would be great to follow structure and flow of int8 decompression example (weights_decompression_matmul) and add additional information about specifics of int4 data storage. If you remember the case that triggered the request for example was related to feeding prepacked weights to oneDNN and dealing with groups and zero-points.

Reply:
Previously in the Teams Pytorch channel, Dmitry provided the following detailed advice:
```
...
Secondly, using this group of 8 along N in case of PT is not required. Group size says how many consecutive points of the tensor zero points are applied to should share a single zero point value. It has nothing to do with how PT pack their zero points.
 
Thirdly, it is the most important HOW these zero points are stored in memory. There was a recent story where IPEX engineer tried to enable oneDNN's int4 and failed to do so because weights were transposed (because of that 8xPack thing), and everything what should have been done was to transpose them again to match oneDNN's API. I would assume this story should follow the same pattern - before calling oneDNN API, it's highly likely those zero points should be transposed and only then passed as an int4 object inside the library to get correct results.
```
@dzarukin What do you suggest? It seems that it's better to pass an int4 object to oneDNN rather than to prepack 8*int4 and pass an int32 object.

Reply:
oneDNN developed API to work with int4 memory objects directly. This hasn't happened to PyTorch yet. Their implementation side has a detail of pre-packing. The example should probably demonstrate how to translate packed 8 int4 values as a single int value language into oneDNN language and what operations should be done in terms of memory (necessary transpositions and/or reorders).